File: src/ashare_evidence/schemas/portfolio.py
"""P o r t f o l i o domain schemas."""
from __future__ import annotations

from ashare_evidence.contract_status import STATUS_PENDING_REBUILD
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field


class TradingRuleCheckView(BaseModel):
    code: str
    title: str
    status: str
    detail: str


class PortfolioHoldingView(BaseModel):
    symbol: str
    name: str
    quantity: int
    avg_cost: float
    last_price: float
    prev_close: float | None = None
    market_value: float
    unrealized_pnl: float
    realized_pnl: float
    total_pnl: float
    holding_pnl_pct: float | None = None
    today_pnl_amount: float
    today_pnl_pct: float | None = None
    portfolio_weight: float
    pnl_contribution: float


class PortfolioAttributionView(BaseModel):
    label: str
    amount: float
    contribution_pct: float
    detail: str


class PortfolioNavPointView(BaseModel):
    trade_date: date
    nav: float
    benchmark_nav: float
    drawdown: float
    exposure: float
    observed_at: datetime | None = None


class PortfolioOrderAuditView(BaseModel):
    order_key: str
    symbol: str
    stock_name: str
    order_source: str
    side: str
    requested_at: datetime
    status: str
    quantity: int
    order_type: str
    avg_fill_price: float | None = None
    gross_amount: float
    checks: list[TradingRuleCheckView] = Field(default_factory=list)


class BenchmarkContextView(BaseModel):
    benchmark_id: str
    benchmark_type: str
    benchmark_symbol: str | None = None
    benchmark_label: str
    source: str
    source_classification: str | None = None
    as_of_time: datetime | None = None
    available_time: datetime | None = None
    status: str = STATUS_PENDING_REBUILD
    note: str | None = None
    artifact_id: str | None = None
    manifest_id: str | None = None
    benchmark_definition: str | None = None


class PortfolioPerformanceView(BaseModel):
    total_return: float
    benchmark_return: float
    excess_return: float
    realized_pnl: float
    unrealized_pnl: float
    fee_total: float
    tax_total: float
    max_drawdown: float
    current_drawdown: float
    order_count: int
    annualized_return: float | None = None
    annualized_excess_return: float | None = None
    sharpe_like_ratio: float | None = None
    turnover: float | None = None
    win_rate_definition: str | None = None
    win_rate: float | None = None
    capacity_note: str | None = None
    artifact_id: str | None = None
    validation_mode: str | None = None
    benchmark_definition: str | None = None
    cost_definition: str | None = None
    cost_source: str | None = None


class ExecutionPolicyView(BaseModel):
    status: str = STATUS_PENDING_REBUILD
    label: str
    summary: str
    policy_type: str | None = None
    source: str | None = None
    note: str | None = None
    constraints: list[str] = Field(default_factory=list)


class PortfolioSummaryView(BaseModel):
    portfolio_key: str
    name: str
    mode: str
    mode_label: str
    strategy_summary: str
    strategy_label: str
    strategy_status: str | None = STATUS_PENDING_REBUILD
    benchmark_symbol: str | None = None
    status: str
    starting_cash: float
    available_cash: float
    market_value: float
    net_asset_value: float
    invested_ratio: float
    total_return: float
    benchmark_return: float
    excess_return: float
    benchmark_status: str | None = STATUS_PENDING_REBUILD
    benchmark_note: str | None = None
    realized_pnl: float
    unrealized_pnl: float
    fee_total: float
    tax_total: float
    max_drawdown: float
    current_drawdown: float
    order_count: int
    active_position_count: int
    rule_pass_rate: float
    recommendation_hit_rate: float | None = None
    market_data_timeframe: str
    last_market_data_at: datetime | None = None
    benchmark_context: BenchmarkContextView
    performance: PortfolioPerformanceView
    execution_policy: ExecutionPolicyView
    validation_status: str = STATUS_PENDING_REBUILD
    validation_note: str | None = None
    validation_artifact_id: str | None = None
    validation_manifest_id: str | None = None
    alerts: list[str] = Field(default_factory=list)
    rules: list[TradingRuleCheckView] = Field(default_factory=list)
    holdings: list[PortfolioHoldingView] = Field(default_factory=list)
    attribution: list[PortfolioAttributionView] = Field(default_factory=list)
    nav_history: list[PortfolioNavPointView] = Field(default_factory=list)
    recent_orders: list[PortfolioOrderAuditView] = Field(default_factory=list)

File: frontend/src/types/portfolio.ts
// portfolio domain types

export interface TradingRuleCheckView {
  code: string;
  title: string;
  status: string;
  detail: string;
}

export interface PortfolioHoldingView {
  symbol: string;
  name: string;
  quantity: number;
  avg_cost: number;
  last_price: number;
  prev_close?: number | null;
  market_value: number;
  unrealized_pnl: number;
  realized_pnl: number;
  total_pnl: number;
  holding_pnl_pct?: number | null;
  today_pnl_amount: number;
  today_pnl_pct?: number | null;
  portfolio_weight: number;
  pnl_contribution: number;
}

export interface PortfolioAttributionView {
  label: string;
  amount: number;
  contribution_pct: number;
  detail: string;
}

export interface PortfolioNavPointView {
  trade_date: string;
  nav: number;
  benchmark_nav: number;
  drawdown: number;
  exposure: number;
  observed_at?: string | null;
}

export interface PortfolioOrderAuditView {
  order_key: string;
  symbol: string;
  stock_name: string;
  order_source: string;
  side: string;
  requested_at: string;
  status: string;
  quantity: number;
  order_type: string;
  avg_fill_price?: number | null;
  gross_amount: number;
  checks: TradingRuleCheckView[];
}

export interface BenchmarkContextView {
  benchmark_id: string;
  benchmark_type: string;
  benchmark_symbol?: string | null;
  benchmark_label: string;
  source: string;
  source_classification?: string | null;
  as_of_time?: string | null;
  available_time?: string | null;
  status: string;
  note?: string | null;
  artifact_id?: string | null;
  manifest_id?: string | null;
  benchmark_definition?: string | null;
}

export interface PortfolioPerformanceView {
  total_return: number;
  benchmark_return: number;
  excess_return: number;
  realized_pnl: number;
  unrealized_pnl: number;
  fee_total: number;
  tax_total: number;
  max_drawdown: number;
  current_drawdown: number;
  order_count: number;
  annualized_return?: number | null;
  annualized_excess_return?: number | null;
  sharpe_like_ratio?: number | null;
  turnover?: number | null;
  win_rate_definition?: string | null;
  win_rate?: number | null;
  capacity_note?: string | null;
  artifact_id?: string | null;
  validation_mode?: string | null;
  benchmark_definition?: string | null;
  cost_definition?: string | null;
  cost_source?: string | null;
}

export interface ExecutionPolicyView {
  status: string;
  label: string;
  summary: string;
  policy_type?: string | null;
  source?: string | null;
  note?: string | null;
  constraints: string[];
}

export interface PortfolioSummaryView {
  portfolio_key: string;
  name: string;
  mode: string;
  mode_label: string;
  strategy_summary: string;
  strategy_label: string;
  strategy_status?: string;
  benchmark_symbol?: string | null;
  status: string;
  starting_cash: number;
  available_cash: number;
  market_value: number;
  net_asset_value: number;
  invested_ratio: number;
  total_return: number;
  benchmark_return: number;
  excess_return: number;
  benchmark_status?: string;
  benchmark_note?: string | null;
  realized_pnl: number;
  unrealized_pnl: number;
  fee_total: number;
  tax_total: number;
  max_drawdown: number;
  current_drawdown: number;
  order_count: number;
  active_position_count: number;
  rule_pass_rate: number;
  recommendation_hit_rate?: number;
  market_data_timeframe: string;
  last_market_data_at?: string | null;
  benchmark_context: BenchmarkContextView;
  performance: PortfolioPerformanceView;
  execution_policy: ExecutionPolicyView;
  validation_status: string;
  validation_note?: string | null;
  validation_artifact_id?: string | null;
  validation_manifest_id?: string | null;
  alerts: string[];
  rules: TradingRuleCheckView[];
  holdings: PortfolioHoldingView[];
  attribution: PortfolioAttributionView[];
  nav_history: PortfolioNavPointView[];
  recent_orders: PortfolioOrderAuditView[];
}

